File: CompressImageByHuffmanExample/src/main/java/com/lishuaihua/test/MainActivity.java
package com.lishuaihua.test;

import android.Manifest;
import android.content.Intent;
import android.content.pm.PackageManager;
import android.content.res.AssetManager;
import android.graphics.Bitmap;
import android.graphics.BitmapFactory;
import android.os.AsyncTask;
import android.os.Bundle;
import android.os.Environment;
import android.support.annotation.NonNull;
import android.support.v4.app.ActivityCompat;
import android.support.v7.app.AppCompatActivity;
import android.text.format.Formatter;
import android.util.Log;
import android.view.View;
import android.widget.ImageView;
import android.widget.Toast;

import com.bumptech.glide.Glide;
import com.lishuaihua.compress.ImageCompress;

import java.io.File;
import java.io.FileInputStream;
import java.io.FileNotFoundException;
import java.io.FileOutputStream;
import java.io.IOException;
import java.io.InputStream;
import java.io.OutputStream;
import java.util.ArrayList;

import me.iwf.photopicker.PhotoPicker;

public class MainActivity extends AppCompatActivity {

    private static final String TAG = "MainActivity";


    //测试图片的存位置
    private String desPath = Environment.getExternalStorageDirectory().getAbsolutePath() + File.separator + "hfresult.jpg";
    private ImageView iv2;
    private ImageView iv1;
    private String path;

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setContentView(R.layout.activity_main);
        iv1 = findViewById(R.id.iv1);
        iv2 = findViewById(R.id.iv2);
        checkPermission();


        findViewById(R.id.btn_compress)
                .setOnClickListener(new View.OnClickListener() {
                    @Override
                    public void onClick(final View v) {
                        compressImage(path,q,50,desPath);

                    }
                });


        findViewById(R.id.btn_select_ptoto)
                .setOnClickListener(new View.OnClickListener() {
                    @Override
                    public void onClick(final View v) {
                        PhotoPicker.builder()
                                .setPhotoCount(9)
                                .setShowCamera(true)
                                .setShowGif(true)
                                .setPreviewEnabled(false)
                                .start(MainActivity.this, PhotoPicker.REQUEST_CODE);
                    }
                });

    }

    int[] q = {10};

    private void compressImage( String path,  int[] quality, int totalSize, String desPath) {
         compressImageByHuffman(path, quality,totalSize, desPath);

    }

    /**
     * 对比压缩出的同等质量的图片，使用哈夫曼算法的话，压缩的更小
     */
    private void compressImageByHuffman(final String path, final int[] quality, final  int totalSize,final String desPath) {
        final float[] size = {0};
        new AsyncTask<Void, Void, Void>() {
            @Override
            protected Void doInBackground(Void... params) {
                File file = new File(path);
                Bitmap bitmap = BitmapFactory.decodeFile(file.getAbsolutePath());
                int i = ImageCompress.nativeCompressBitmap(bitmap, quality[0], desPath, true);
                return null;
            }

            @Override
            protected void onPostExecute(Void aVoid) {

                Toast.makeText(MainActivity.this, "压缩完成", Toast.LENGTH_LONG).show();
                File file1 = new File(desPath);
                Glide.with(MainActivity.this).load(file1).into(iv2);
                String fileSize = Formatter.formatFileSize(MainActivity.this, getFileSize(file1));
                Log.i(TAG, "图片压缩后大小：" + fileSize);
                size[0] = (float) new File(desPath).length() / 1024;
                while (quality[0]>0&&size[0]>totalSize){
                    quality[0]=quality[0]/2;
                    Log.i(TAG,"quality="+quality[0]);
                     compressImageByHuffman(path, quality,totalSize, desPath);

                }

            }
        }.execute();

    }

    /**
     * 6.0 权限申请
     */
    private void checkPermission() {
        if (checkCallingOrSelfPermission(Manifest.permission.WRITE_EXTERNAL_STORAGE)
                != PackageManager.PERMISSION_GRANTED) {
            ActivityCompat.requestPermissions(this, new String[]{Manifest.permission
                    .WRITE_EXTERNAL_STORAGE}, 100);
        }
    }

    @Override
    protected void onActivityResult(int requestCode, int resultCode, Intent data) {
        super.onActivityResult(requestCode, resultCode, data);

        if (resultCode == RESULT_OK && requestCode == PhotoPicker.REQUEST_CODE) {
            if (data != null) {
                ArrayList<String> photos = data.getStringArrayListExtra(PhotoPicker.KEY_SELECTED_PHOTOS);
                path = photos.get(0);
                File file = new File(path);
                Log.i(TAG, "原图片大小：" + Formatter.formatFileSize(MainActivity.this, getFileSize(file)));
                Glide.with(MainActivity.this).load(file).into(iv1);
            }
        }
    }


    /**
     * 获取指定文件大小
     *
     * @param file
     * @return
     * @throws Exception
     */
    public static long getFileSize(File file) {
        long size = 0;
        if (file.exists()) {
            FileInputStream fis = null;
            try {
                fis = new FileInputStream(file);
                size = fis.available();

            } catch (IOException e) {
                e.printStackTrace();
            }
        } else {
            Log.e("获取文件大小", "文件不存在!");
        }
        return size;
    }
}

File: app/src/main/java/com/lishuaihua/test/MainActivity.kt
package com.lishuaihua.test

import android.Manifest
import android.content.Intent
import android.os.Bundle
import android.text.format.Formatter
import android.util.Log
import android.view.View
import android.widget.ImageView
import android.widget.TextView
import android.widget.Toast
import com.bumptech.glide.Glide
import com.hjq.permissions.OnPermission
import com.hjq.permissions.Permission
import com.hjq.permissions.XXPermissions
import com.lishuaihua.compress.BaseViewModel
import com.lishuaihua.compress.CompressListener
import com.lishuaihua.imageselector.utils.ImageSelector
import java.io.File
import java.io.FileInputStream
import java.io.IOException
import java.util.*

class MainActivity : BaseActivity<BaseViewModel>() {
    var permissions =
        arrayOf(Manifest.permission.CAMERA, Manifest.permission.MANAGE_EXTERNAL_STORAGE)

    //测试图片的存位置
    private lateinit var desPath: String
    private lateinit var iv2: ImageView
    private lateinit var iv1: ImageView
    private lateinit var tvCompressedImg: TextView
    private lateinit var tvOriginImg: TextView
    private var path: String? = null
    override fun initView(savedInstanceState: Bundle?) {
        super.initView(savedInstanceState)
        iv1 = findViewById(R.id.iv_origin_img)
        tvOriginImg = findViewById(R.id.tv_origin_img)
        iv2 = findViewById(R.id.iv_compressed_img)
        tvCompressedImg = findViewById(R.id.tv_compressed_img)

        checkPermission()

        findViewById<View>(R.id.btn_compress)
            .setOnClickListener {
                compressImage()
            }

        findViewById<View>(R.id.btn_select_photo)
            .setOnClickListener {
                ImageSelector.builder()
                    .useCamera(false) // 设置是否使用拍照 需求文档v2 p38 点击上传图标则打开本机相册
                    .setSingle(true) //设置是否单选
                    .setMaxSelectCount(1) // 图片的最大选择数量，小于等于0时，不限数量。
                    .start(this@MainActivity, Constants.PICK_IMAGE) // 打开相册
            }
        iv1.setOnClickListener {
            if (!StringUtil.isEmpty(path)){
                val imgUrls = ArrayList<String>()
                imgUrls.add(path!!)
                val intent = Intent()
                val bundle = Bundle()
                bundle.putStringArrayList("URLS", imgUrls)
                bundle.putInt("POS", 0)
                intent.putExtras(bundle)
                intent.setClass(this@MainActivity, ImagePreviewActivity::class.java)
              startActivity(intent)
            }
        }
        iv2.setOnClickListener {
            if (!StringUtil.isEmpty(desPath)){
                val imgUrls = ArrayList<String>()
                imgUrls.add(desPath!!)
                val intent = Intent()
                val bundle = Bundle()
                bundle.putStringArrayList("URLS", imgUrls)
                bundle.putInt("POS", 0)
                intent.putExtras(bundle)
                intent.setClass(this@MainActivity, ImagePreviewActivity::class.java)
                startActivity(intent)
            }
        }
    }

    private fun compressImage() {
        desPath = cacheDir.toString() + System.currentTimeMillis() + ".jpg"
        vm.compressImage(
            path = path,
            quality = intArrayOf(100),
            totalSize = 1024,
            desPath = desPath,
            listener = object : CompressListener {
                override fun startCompress() {
                    showLoading("Compressing")
                }

                override fun completedCompress() {
                    hideLoading()
                    var msg = "压缩后的图: " + Formatter.formatFileSize(
                        this@MainActivity,
                        getFileSize(File(desPath))
                    )
                    tvCompressedImg.text = msg.orEmpty()
                    Glide.with(this@MainActivity).load(File(desPath)).into(iv2)
                }
            })
    }


    /**
     * 6-11权限申请
     */
    private fun checkPermission() {
        XXPermissions.with(this).permission(permissions).request(object : OnPermission {
            override fun hasPermission(granted: MutableList<String>?, all: Boolean) {
                if (all) {
                    Toast.makeText(this@MainActivity, "获取相机和读写权限成功", Toast.LENGTH_LONG).show()
                } else {
                    Toast.makeText(this@MainActivity, "获取部分权限成功，但部分权限未正常授予", Toast.LENGTH_LONG)
                        .show()
                }
            }

            override fun noPermission(denied: MutableList<String>?, never: Boolean) {
                if (never) {
                    // 如果是被永久拒绝就跳转到应用权限系统设置页面
                    Toast.makeText(this@MainActivity, "被永久拒绝授权，请手动授予录音和日历权限", Toast.LENGTH_LONG)
                        .show()
                    XXPermissions.startPermissionActivity(
                        this@MainActivity,
                        permissions.toMutableList()
                    )
                } else {
                    Toast.makeText(this@MainActivity, "获取相机和读写权限失败", Toast.LENGTH_LONG).show()
                }
            }

        })
    }

    override fun onActivityResult(requestCode: Int, resultCode: Int, data: Intent?) {
        super.onActivityResult(requestCode, resultCode, data)
        if (resultCode == RESULT_OK && requestCode == Constants.PICK_IMAGE) {
            if (data != null) {
                val photos = data.getStringArrayListExtra(ImageSelector.SELECT_RESULT)
                path = photos!![0]
                val file = File(path)
                var msg = "压缩前的图：" + Formatter.formatFileSize(this@MainActivity, getFileSize(file))
                tvOriginImg.text = msg.orEmpty()
                Glide.with(this@MainActivity).load(file).into(iv1!!)
            }
        }

    }

    companion object {
        private const val TAG = "MainActivity"

        /**
         * 获取指定文件大小
         *
         * @param file
         * @return
         * @throws Exception
         */
        fun getFileSize(file: File): Long {
            var size: Long = 0
            if (file.exists()) {
                Log.e(TAG, file.absolutePath)
                var fis = FileInputStream(file)
                try {
                    size = fis.available().toLong()
                } catch (e: IOException) {
                    Log.e(TAG, e.message!!)
                }
            } else {
                Log.e("获取文件大小", "文件不存在!")
            }
            return size
        }
    }

    override fun getLayoutId(): Int = R.layout.activity_main
}